### PR TITLE
add promxial policy optimization template to pl_examples 

### LINKED ISSUE #5393: adding proximal policy optimization template to pl_examples
## 🚀 Feature
An implementation of proximal policy optimization (PPO) to pl_examples. 

### Motivation
pl_examples features one reinforcement learning algorithm- DQN. It could be good to have a template of a new policy gradient method. 

I had [implemented](https://github.com/sid-sundrani/ppo_lightning) PPO in Lightning and it was mentioned on Slack that this could be useful to add to pl_examples

### PR DESCRIPTION
**Pull request by sid-sundrani**
## What does this PR do?
Adds example of implementation of proximal policy optimization in PyTorch Lightning. 

I had implemented PPO in Lightning in my repo and there are some results [there](https://github.com/sid-sundrani/ppo_lightning) as well. If it's good to have an example for PPO in pl_examples, I'm happy to help add it 

Fixes #5393

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by sid-sundrani** [comment]
I added my code to the wrong branch (master) and I'm merging with dev.. Not sure what to do, there seem to be 448 commits now in this PR.. Should I close this PR and add my code to PyTorchLightning:release/1.0.x and create a PR again?

**Comment by Borda** [comment]
@sid-sundrani well in fact the correct for you is `release/1.2-dev` I have fixed it for you 🐰

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        """
+        rews = rewards + [last_value]
+        vals = values + [last_value]
+        # GAE
+        delta = [rews[i] + self.gamma * vals[i + 1] - vals[i] for i in range(len(rews) - 1)]
+        adv = self.discount_rewards(delta, self.gamma * self.lam)
+
+        return adv
+
+    def train_batch(
```
[the remark this answers is not in the document]
Comment (sid-sundrani) [reply pl_examples/domain_templates/reinforce_learn_ppo.py:281]: Actually, yes it would make more sense. `train_batch` seems a little ambiguous here

I was trying to be consistent with how it is named in the [Reinforce](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/8a4cf8f61644c28d6df54ccffe3a52d6f5fce5a6/pl_bolts/models/rl/reinforce_model.py#L165) RL implementation in Lightning Bolts.

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    trainer = pl.Trainer(max_epochs=70)
+    trainer.fit(model)
+
+
+if __name__ == '__main__':
+    cli_lightning_logo()
+    torch.manual_seed(0)
+    np.random.seed(0)
+
+    parser = argparse.ArgumentParser(add_help=False)
```
[the remark this answers is not in the document]
Comment (sid-sundrani) [reply pl_examples/domain_templates/reinforce_learn_ppo.py:469]: I don't think it works with DDP- but I'm saying that only because all the RL [implementations](https://github.com/PyTorchLightning/pytorch-lightning-bolts/tree/master/pl_bolts/models/rl) in lightning bolts don't support it (says in their comments). And this PPO implementation has very similar structure and data generation as [Reinforce](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/8a4cf8f61644c28d6df54ccffe3a52d6f5fce5a6/pl_bolts/models/rl/reinforce_model.py#L165)

An iterable dataset is generated before every epoch by `train_batch` method, which performs trajectory collection using the model

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+    parser = PPOLightning.add_model_specific_args(parser)
```
[the remark this answers is not in the document]
Comment (sid-sundrani) [reply pl_examples/domain_templates/reinforce_learn_ppo.py]: Sure like this?..

**Review by ananyahjha93 — APPROVED** [review]

**Review by teddykoker — APPROVED** [review]
LGTM

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        self.lam = lam
+        self.max_episode_len = max_episode_len
+        self.clip_ratio = clip_ratio
+        self.save_hyperparameters()
+
+        self.env = gym.make(env)
+        # value network
+        self.critic = create_mlp(self.env.observation_space.shape, 1)
+        # policy network (agent)
+        if type(self.env.action_space) == gym.spaces.box.Box:
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:207]: I recommend to use below code pattern
```python
if isinstance(self.env.action_space, gym.spaces.box.Box):
```

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+            act_dim = self.env.action_space.shape[0]
+            actor_mlp = create_mlp(self.env.observation_space.shape, act_dim)
+            self.actor = ActorContinous(actor_mlp, act_dim)
+        elif type(self.env.action_space) == gym.spaces.discrete.Discrete:
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:211]: here too

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+            actor_mlp = create_mlp(self.env.observation_space.shape, self.env.action_space.n)
+            self.actor = ActorCategorical(actor_mlp)
+        else:
+            raise NotImplementedError('Env action space should be of type Box (continous) or Discrete (categorical)'
+                                      'Got type: ', type(self.env.action_space))
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:216]: Use f-string 😄 
```python
raise NotImplementedError(f'Env action space should be of type Box (continous) or Discrete (categorical) Got type: {type(self.env.action_space)}')
```

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+                # logging
+                self.avg_reward = sum(self.epoch_rewards) / self.steps_per_epoch
+
+                # if epoch ended upbruptly, exlude last cut-short episode to prevent stats skewness
+                if not done:
+                    total_epoch_reward = sum(self.epoch_rewards[:-1])
+                    nb_episodes = len(self.epoch_rewards) - 1
+                else:
+                    total_epoch_reward = sum(self.epoch_rewards)
+                    nb_episodes = len(self.epoch_rewards)
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:362]: There is room for deduplication in this code.
like
```python
epoch_rewards = self.epoch_rewards
if not done:
    epoch_rewards = epoch_rewards[:-1]
total_epoch_reward = sum(epoch_rewards)
nb_episodes = len(epoch_rewards)
```

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+
+        return optimizer_actor, optimizer_critic
+
+    def optimizer_step(self, *args, **kwargs):
+        """
+        Run 'nb_optim_iters' number of iterations of gradient descent on actor and critic
+        for each data sample.
+        """
+        for i in range(self.nb_optim_iters):
+            super().optimizer_step(*args, **kwargs)
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:427]: `i` is not used

```python
for _ range(self.nb_optim_iters):
    super().optimizer_step(*args, **kwargs)
```

File: pl_examples/domain_templates/reinforce_learn_ppo.py [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+    model = PPOLightning(**vars(args))
+
+    trainer = pl.Trainer.from_argparse_args(args)
+    trainer.fit(model)
+
+
+if __name__ == '__main__':
+    cli_lightning_logo()
+    torch.manual_seed(0)
+    np.random.seed(0)
```
Comment (marload) [inline pl_examples/domain_templates/reinforce_learn_ppo.py:469]: In pytorchlightning, you can use `pl.seed_everything(0)` to fix the seed.

**Review by marload — CHANGES_REQUESTED** [review]
Awesome!
But I have some request change for code quality. 😄

File: pl_examples/domain_templates/reinforce_learn_ppo.py (hunk 7 above)
Comment (sid-sundrani) [reply pl_examples/domain_templates/reinforce_learn_ppo.py]: Yup this is better

**Review by marload — APPROVED** [review]
LGTM

**Review by tchaton — APPROVED** [review]
